File: src/backend/Goblin.Narfu/Models/Lesson.cs
using System;

namespace Goblin.Narfu.Models;

public record Lesson
{
    public string Id { get; set; }

    public string Type { get; set; }
    public string Name { get; set; }
    public DateTime StartTime { get; set; }
    public DateTime EndTime { get; set; }
    public string StartEndTime { get; set; }
    public int Number { get; set; }
    public string Address { get; set; }
    public string Auditory { get; set; }
    public string Teacher { get; set; }
    public string Groups { get; set; }

    public string Link { get; set; }

    public bool IsExam => LessonType == LessonType.Exam;

    public LessonType LessonType => GetLessonType();

    private LessonType GetLessonType()
    {
        if(Type.Contains("экзамен", StringComparison.InvariantCultureIgnoreCase) ||
           Type.Contains("зачет", StringComparison.InvariantCultureIgnoreCase))
        {
            return LessonType.Exam;
        }

        if(Type.Contains("практическ", StringComparison.InvariantCultureIgnoreCase))
        {
            return LessonType.Practical;
        }

        if(Type.Contains("лабораторн", StringComparison.InvariantCultureIgnoreCase))
        {
            return LessonType.Laboratory;
        }
        
        if(Type.Contains("лекции", StringComparison.InvariantCultureIgnoreCase))
        {
            return LessonType.Lecture;
        }
        
        if(Type.Contains("консультация", StringComparison.InvariantCultureIgnoreCase))
        {
            return LessonType.Consultation;
        }

        return LessonType.Unknown;
    }
}

File: src/frontend/functions/services/lesson.ts
import {LessonType} from "./lesson_type";

export class Lesson {
    id: string;
    type: string;
    name: string;
    startTime: Date;
    endTime: Date;
    startEndTime: string;
    number: number;
    address: string;
    auditory: string;
    teacher: string;
    groups: string;
    lessonType: LessonType;

    link: string | undefined;

    constructor(id: string, type: string, name: string, start: Date, end: Date, startend: string,
                number: number, address: string, aud: string, teacher: string, group: string, link: string | undefined) {
        this.id = id;
        this.type = type;
        this.name = name.replace('. ', '.').replace('.', '. ');
        this.startTime = start;
        this.endTime = end;
        this.startEndTime = startend;
        this.number = number;
        this.address = address;
        this.auditory = aud;
        this.teacher = teacher;
        this.groups = group;
        this.lessonType = this.getLessonType(type);
        this.link = link;
    }
    
    getLessonType(str: string): LessonType {
        str = str.toLowerCase();
        if(str.includes("экзамен") || str.includes("зачет")) {
            return LessonType.exam;
        }
        if(str.includes("практическ")) {
            return LessonType.practical;
        }
        if(str.includes("лабораторн")) {
            return LessonType.laboratory;
        }
        if(str.includes("лекции")) {
            return LessonType.lecture;
        }
        if(str.includes("консультация")) {
            return LessonType.consultation;
        }
        
        return LessonType.unknown;
    }
}